src/System.Text.Utf8String/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,782 @@
+            }
+
+            if (s == string.Empty)
+            {
+                _buffer = ReadOnlySpan<byte>.Empty;
+            }
+            else
+            {
+                _buffer = new ReadOnlySpan<byte>(GetUtf8BytesFromString(s));
+            }
+        }
+
+        /// <summary>
+        /// This constructor is for use by the compiler.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Utf8Span(RuntimeFieldHandle utf8Data, int length) : this(CreateArrayFromFieldHandle(utf8Data, length))
+        {
+        }
+
+        public static explicit operator Utf8Span(ArraySegment<byte> utf8Bytes)
+        {
+            return new Utf8Span(utf8Bytes);
+        }
+
+        static byte[] CreateArrayFromFieldHandle(RuntimeFieldHandle utf8Data, int length)
+        {
+            var array = new byte[length];
+            RuntimeHelpers.InitializeArray(array, utf8Data);
+            return array;
+        }
+
+        public static Utf8Span Empty { get { return s_empty; } }
+
+        /// <summary>
+        /// Returns length of the string in UTF-8 code units (bytes)
+        /// </summary>
+        public int Length
+        {
+            get
+            {
+                return _buffer.Length;
+            }
+        }
+
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(_buffer);
+        }
+
+        public CodePointEnumerable CodePoints
+        {
+            get
+            {
+                return new CodePointEnumerable(_buffer);
+            }
+        }
+
+        public byte this[int i]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                // there is no need to check the boundaries -> Span is going to do this on it's own
+                return (byte)_buffer[i];
+            }
+        }
+
+        public static implicit operator ReadOnlySpan<byte>(Utf8Span utf8)
+        {
+            return utf8.Bytes;
+        }
+
+        public static explicit operator Utf8Span(string s)
+        {
+            return new Utf8Span(s);
+        }
+
+        public static explicit operator string(Utf8Span s)
+        {
+            return s.ToString();
+        }
+
+        public ReadOnlySpan<byte> Bytes => _buffer;
+
+        public override string ToString()
+        {
+            var status = Encodings.Utf8.ToUtf16Length(this.Bytes, out int needed);
+            if (status != Buffers.OperationStatus.Done)
+                return string.Empty;
+
+            // UTF-16 is 2 bytes per char
+            var chars = new char[needed >> 1];
+            var utf16 = new Span<char>(chars).AsBytes();
+            status = Encodings.Utf8.ToUtf16(this.Bytes, utf16, out int consumed, out int written);
+            if (status != Buffers.OperationStatus.Done)
+                return string.Empty;
+
+            return new string(chars);
```

Comment:
The way this method should be implemented is similar to how Encodings.Ascii.ToUtf16String is implemented, i.e. you should compute how many bytes would need, then create System.String of appropriate length, then pin the string, and then write the transcoded bytes directly to the pinned string.

Reply:
Could you add a TODO in the source to optimize it later? Thanks.